feat(aio): implement ScrollSpy service (to highlight the active section in ToC)

gkalpak opened the pull request:
Seems to work nicely, but no tests yet 😳
It turned out way more complicated than I wanted t to be. Alternative ideas or suggestions welcome 😕

gkalpak committed:
feat(aio): implement ScrollSpy service (to highlight the active section in ToC)

petebacondarwin commented on aio/src/app/shared/scroll-spy.service.ts:
```diff
@@ -0,0 +1,130 @@
+
+    return {
+      active: spiedGroup.activeScrollItem.asObservable(),
+      unspy: () => this.unspy(spiedGroup)
```
Given that observables are inherently unsubscribable, it would seem that this `unspy` is unnecessary.
Why not put the logic of `ScrollSpyService.unspy` into the logic to handle the observable being unsubscribed from?

gkalpak replied:
I think, because there is no guarantee that the consumer will actually subscribe to `active`.

petebacondarwin replied:
which indicates to me that we are doing too much work upfront and not allowing our observables to be lazy enough :-)

gkalpak replied:
I am actually very tempted to re-architecture `ScrollSpy`, so that it doesn't  keep track of the `ScrollSpiedElementGroup`s and `spyOn` would just return a cold, shareable observable (with access to shared, ref-counting `onResize`/`onScroll` observables). But I think I'll pass :smiley:

I will add it to a list of possible future enhancements/improvements for scroll-spying, though.

mary-poppins commented:
The angular.io preview for 30f99dc8aa9422d3a26a29dbe367b920c1afe055 is available [here][1].

[1]: https://pr16703-30f99dc8aa9422d3a26a29dbe367b920c1afe055.ngbuilds.io/

wardbell commented on aio/src/testing/scroll-spy.service.ts:
```diff
@@ -0,0 +1,12 @@
+import { Subject } from 'rxjs/Subject';
+import { ScrollItem, ScrollSpyInfo } from 'app/shared/scroll-spy.service';
+
+export class MockScrollSpyService {
```
Why create a `MockScrollSpyService` way out here?

IMO, following DAMP principles, we should create mock/test classes where they are used so that the reader isn't wandering around the code base. They shouldn't be in far away files unless ...

1. they are reused significantly
2. they have non-trivial implementations 

This is a trivial mock class, referenced (but not yet used!) in one spec file only and unlikely to be reused. YAGNI?

I could not, in good conscience, block this PR for this objection, particularly as we have decided to postpone testing style discussions until after our release.

gkalpak replied:
Fair enough :smiley: If this turns out to only be used in one place I will move it into the spec file.

wardbell commented on aio/src/app/shared/toc.service.spec.ts:
```diff
@@ -20,6 +22,7 @@ describe('TocService', () => {
     injector = ReflectiveInjector.resolveAndCreate([
       { provide: DomSanitizer, useClass: TestDomSanitizer },
       { provide: DOCUMENT, useValue: document },
+      { provide: ScrollSpyService, useValue: new MockScrollSpyService() },
```
Why `useValue: new MSSS()` instead of `useClass: MSSS` ?

wardbell commented on aio/src/app/shared/toc.service.ts:
```diff
@@ -61,6 +73,13 @@ export class TocService {
+    const headings = docElement.querySelectorAll('h2,h3');
+    const skipNoTocHeadings = (heading: HTMLHeadingElement) => !/(?:no-toc|notoc)/i.test(heading.className);
+
+    return Array.prototype.filter.call(headings, skipNoTocHeadings);
```
I don't really mind a for loop but I see you do ;-)

wardbell reviewed (changes requested):
I have my snowflake quibbles.

The significant issue is that the TOC does not scroll when the selected TOC item is not visible.

Look at the Routing guide to see this problem. As you scroll through that long guide, eventually the current TOC item is no longer visible on the right side TOC.

I suggest that when the current TOC item is no longer visible, the right-side TOC should scroll the current TOC item into view, possibly the middle of the TOC panel.

Thoughts?

mary-poppins commented:
The angular.io preview for 0a97dfc950996406d0cfd31a923ed4223c3c4cbd is available [here][1].

[1]: https://pr16703-0a97dfc950996406d0cfd31a923ed4223c3c4cbd.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 98e22145db209aaaca5bd9886eeaa1ea52790e22 is available [here][1].

[1]: https://pr16703-98e22145db209aaaca5bd9886eeaa1ea52790e22.ngbuilds.io/

petebacondarwin commented on aio/src/app/embedded/toc/toc.component.ts:
```diff
@@ -47,6 +52,34 @@ export class TocComponent implements OnInit, OnDestroy {
+  ngAfterViewInit() {
+    if (!this.isEmbedded) {
+      this.tocService.activeItemIndex
+          .takeUntil(this.onDestroy)
```
totally irrelevant here but I think a better convention would be to rename `this.onDestroy` to `this.destroyEvent`. This is because `onDestroy` sounds like a handler for the destroy event but actually it is an event source (i.e. an `Observable`).

petebacondarwin commented on aio/src/app/embedded/toc/toc.component.ts:
```diff
@@ -47,6 +52,34 @@ export class TocComponent implements OnInit, OnDestroy {
+      Observable.combineLatest(this.items.changes, this.tocService.activeItemIndex)
+          .takeUntil(this.onDestroy)
+          .subscribe(([items, index]) => {
+            if (index === null || index >= items.length) {
```
Why not set `this.activeItemIndex` in this subscribe, rather than creating an additional one above?

gkalpak replied:
I tried hard to do it like that, but I was getting a error that some value changed after it had been checked (I think it was the `i === activeIndex` in the view).

I don't know much about the ViewChildren and change-detection, so I amy be missing something. (Suggestions welcome :smiley:)

mary-poppins commented:
The angular.io preview for bdb8face4c49c0f716af288561bd13d7e37e30e6 is available [here][1].

[1]: https://pr16703-bdb8face4c49c0f716af288561bd13d7e37e30e6.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 26fe492f7a170bbb9659f6fef0b66829d37f8d59 is available [here][1].

[1]: https://pr16703-26fe492f7a170bbb9659f6fef0b66829d37f8d59.ngbuilds.io/

gkalpak commented:
This is now ready for review. I also added the scrolling of the selected ToC item into view (if it is not visible).
The unit tests for `ScrollSpyService` are not very fine-grained, but I felt that putting more effort/time into unit/e2e testing (for this non-critical feature) wasn't worth it at the moment.

mary-poppins commented:
The angular.io preview for 7578b998826a1b031b587fdcc047e677df1a86a9 is available [here][1].

[1]: https://pr16703-7578b998826a1b031b587fdcc047e677df1a86a9.ngbuilds.io/

sjtrimble commented:
@gkalpak Is the "active" class assigned to the a tag the one that we'll use for styling? If so, can we make it applicable to the li element instead so I can add the circle before and then dig down into the a tag for styling as well? (I have made the design refresh css code work this way right now)

... OR is that class specifically on the a tag being used for another reason that I'm missing?

gkalpak commented:
> Is the "active" class assigned to the a tag the one that we'll use for styling? If so, can we make it applicable to the li element instead?

Yes and sure.

> so I can add the circle before

I thought there would be no circles 😈

##
I'll move the class.

petebacondarwin commented on aio/src/app/embedded/toc/toc.component.spec.ts:
```diff
@@ -258,6 +260,161 @@ describe('TocComponent', () => {
+        page = setPage();
+        const newActive = page.anchorItems.find(By.css('.active')).nativeElement;
+
+        expect(newActive).not.toBe(oldActive);
```
Doesn't this behaviour depend upon whether Angular reuses DOM nodes in `ngFor` or not?

gkalpak replied:
Good point 👍

petebacondarwin replied:
I wondered if this test should just insert a new item at the start of the array, rather than replacing it altogether? That would mean that the nodes might just shift and not be replaced but the component would need to move the active class to a different node?

gkalpak replied:
I like it. Let's do both :smiley:

gkalpak replied:
Done.

petebacondarwin commented on aio/src/app/shared/scroll-spy.service.ts:
```diff
@@ -0,0 +1,171 @@
+    this.spiedElements.sort((a, b) => b.top - a.top);   // Sort in descending `top` order.
+  }
+
+  needsRecalibration(scrollTop: number, topOffset: number) {
```
A comment or two in this class to explain the goal and approach of "(re)calibration" would save time in the future when people come to maintain this code.

gkalpak replied:
While adding comments, I realized that this check is not necessary. The reason my original (less expensive) check didn't work was a bug: I was checking for changes in the viewport height (`body.clientHeight`), while I should be checking for changes in the total content height (`body.scrollHeight`).

<sub>(I 'll still add comments to the rest of methods :wink:)</sub>

petebacondarwin commented on aio/src/app/shared/scroll-spy.service.ts:
```diff
@@ -0,0 +1,171 @@
+  spyOn(elements: Element[]): ScrollSpyInfo {
+    if (!this.spiedElementGroups.length) {
+      this.resizeEvents.subscribe(() => this.onResize());
+      this.scrollEvents.subscribe(() => this.onScroll());
```
The way this service is architected, it looks like we might be able to spy on multiple sets of `elements` at once. If that is the case, then it looks to me like we would be creating unnecessary additional subscriptions to the `resizeEvents` and `scrollEvents` observables for the 2nd and subsequent elements sets.

Given that right now we don't spy on more than one set at a time, this doesn't matter. But if we did then we would need a more clever subscribe/unsubscribe arrangement (e.g. through `.share()`, which does a `.refCount()` if I remember correctly).

petebacondarwin replied:
Actually, in the case of spying on multiple sets of elements, we would probably want to specify the scroll top (i.e. the scroll container dimensions) separately for each set.

gkalpak replied:
Yes, we can spy on multiple sets of elements, but there is only one subscription for each `resize` and `scroll` active at time. Note, we only subscribe when the first element set is spied on and unsubscribe when the last one is "unspied".

petebacondarwin replied:
And given that we are not planning on doing any of that, we could probably make this service simpler.

petebacondarwin replied:
Oh yeah, I didn't notice the `if (!this.spiedElementGroups.length) {`

petebacondarwin reviewed (approved):
A few Qs but nothing to block the merge.

petebacondarwin commented:
Back to you @gkalpak - please fix up the changes for @sjtrimble, squash and apply the merge label.

sjtrimble commented:
@gkalpak circle will only appear on active item for clear indication of current location. Regardless, it gives us more flexibility for styling purposes.

mary-poppins commented:
The angular.io preview for c333f8731f53edf01dc6d59c45ebb2f91d685c98 is available [here][1].

[1]: https://pr16703-c333f8731f53edf01dc6d59c45ebb2f91d685c98.ngbuilds.io/

gkalpak commented:
Rebased, addressed feedback and simplified a bit (originally in [c333f87][1], now squashed into the first commit). Adding the `merge` label and I'll create an issue to track possible future improvements.

Thx for the reviews!

[1]: https://github.com/angular/angular/pull/16703/commits/c333f8731f53edf01dc6d59c45ebb2f91d685c98

mary-poppins commented:
The angular.io preview for 18e66ce169aecf66ecf75e59593ae6da5277bfe4 is available [here][1].

[1]: https://pr16703-18e66ce169aecf66ecf75e59593ae6da5277bfe4.ngbuilds.io/
